### PR TITLE
OptimizationApplication

### PR DESCRIPTION
**Pull request by RezaNajian**
**📝 Description**
This PR implements a general solver application for FEM-based simulation and optimization. It can handle shape, thickness, topology, and material optimizations. Concurrent multi-scale optimization is also realized. To summarize, the following features characterize OptimizationApplication:

- efficient and consistent filtering techniques for parametrization-free shape, thickness, and material/topology optimization;
- an adaptive gradient-projection technique developed especially for problems with an arbitrarily large number of design variables of different scales;
- modular implementation, which enables analysis and optimization of multi-scale and multi-physics problems.

### CODE REVIEW

File: kratos_salome_plugin.log [hunk 1]
```diff
@@ -0,0 +1,997 @@
+[2022-04-23 16:31:28] [INFO    ] KRATOS SALOME PLUGIN : Plugin version: 1.0.0
```
Comment (loumalouomega) [inline kratos_salome_plugin.log]: Remove this file

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Blocking

File: kratos/includes/variables.h [hunk 2]
```diff
@@ -97,6 +97,8 @@ namespace Kratos
     KRATOS_DEFINE_VARIABLE( double, POSITIVE_FACE_PRESSURE )
     KRATOS_DEFINE_VARIABLE( double, FACE_HEAT_FLUX )
 
+    KRATOS_DEFINE_VARIABLE( double, PD )
```
Comment (loumalouomega) [inline kratos/includes/variables.h]: A more descriptive name?

**Review by roigcarlo — COMMENTED** [review]
Hi @RezaNajian 

Please, consider adding a file descriptor for your app like https://github.com/KratosMultiphysics/Kratos/blob/master/applications/FluidDynamicsApplication/FluidDynamicsApplication.json. Otherwise we won't be able to integrate it in our pipelines.

**Comment by matekelemen** [comment]
image
Umm, can you please break this up into smaller PRs? I doubt there's anyone on this planet capable of reviewing this as it is right now.

**Comment by loumalouomega** [comment]
> > image
> 
> Umm, can you please break this up into smaller PRs? I doubt there's anyone on this planet capable of reviewing this as it is right now.

image

**Comment by sunethwarna** [comment]
Hi guys.... This is not for merging. We will put separate small PRs for specific features in future. This is a working draft PR to collaborate and communicate with each other. :) :)

File: kratos_salome_plugin.log (hunk 1 above)
Comment (RezaNajian) [reply kratos_salome_plugin.log]: Tnx, done

File: kratos/includes/variables.h (hunk 2 above)
Comment (RezaNajian) [reply kratos/includes/variables.h]: removed and moved to OptimizationApplication

**Comment by sunethwarna** [comment]
Hi, this is a stand-alone research application which does not influence any other applications or the core. Adaptation of the code to newer standards and/or Kratos core elements will take place after merging this PR. This application is proven to be working at the current state, and we think it would be easier to make the improvements/optimizations/tuning of the application once it is in the master. Hence I am approving this. I, @RezaNajian (TUM) and Fabian Meister (from TUBs) will be the main responsible persons for this application and for the further improvements.

@loumalouomega We removed the conflicts. Could you please remove the block?

**Review by sunethwarna — APPROVED** [review]

**Review by loumalouomega — APPROVED** [review]

**Pull request merged**